7zip/Makefile

```diff
@@ -1,51 +1,97 @@
-               CPP/7zip/Common/InBuffer.cpp \
-               CPP/7zip/Common/OutBuffer.cpp \
-               CPP/7zip/Common/StreamUtils.cpp \
-               CPP/7zip/Compress/BitlDecoder.cpp \
-               CPP/7zip/Compress/DeflateDecoder.cpp \
-               CPP/7zip/Compress/DeflateEncoder.cpp \
-               CPP/7zip/Compress/LzOutWindow.cpp \
-               CPP/7zip/Compress/ZlibDecoder.cpp \
-               CPP/7zip/Compress/ZlibEncoder.cpp \
-               CPP/Common/MyString.cpp \
-               CPP/Common/StringConvert.cpp \
-               CPP/Common/StringToInt.cpp \
-               CPP/NotWindows/FileIO.cpp \
-               CPP/NotWindows/MyWindows.cpp \
-               CPP/Windows/PropVariant.cpp \
-               deflate7z.cpp
-7ZIP_CXX_OBJ = $(7ZIP_CXX_SRC:.cpp=.o)
-7ZIP_C_SRC =   C/Alloc.c \
-               C/HuffEnc.c \
-               C/LzFind.c \
-               C/Sort.c
-7ZIP_C_OBJ =   $(7ZIP_C_SRC:.c=.o)
-
-%.o: %.cpp
-	$(CXX) -c $(SRC_CXXFLAGS) $(CXXFLAGS) -o $@ $<
-
-%.o: %.c
+OBJDIR := objs
+BLDDIR := $(OBJDIR)/7zip
+MKDIRS := \
+	$(BLDDIR)/CPP/7zip/Archive/Common \
+	$(BLDDIR)/CPP/7zip/Common \
+	$(BLDDIR)/CPP/7zip/Compress \
+	$(BLDDIR)/CPP/Common \
+	$(BLDDIR)/CPP/NotWindows \
+	$(BLDDIR)/CPP/Windows \
+	$(BLDDIR)/C
+
+7ZIP_ARCHIVE_DIR := $(SRCDIR)/CPP/7zip/Archive
+7ZIP_COMMON_DIR := $(SRCDIR)/CPP/7zip/Common
+7ZIP_COMPRESS_DIR := $(SRCDIR)/CPP/7zip/Compress
+
+7ZIP_CXX_SRC := $(7ZIP_ARCHIVE_DIR)/Common/ParseProperties.cpp \
+                $(7ZIP_ARCHIVE_DIR)/DeflateProps.cpp \
+                $(7ZIP_COMMON_DIR)/CWrappers.cpp \
+                $(7ZIP_COMMON_DIR)/FileStreams.cpp \
+                $(7ZIP_COMMON_DIR)/InBuffer.cpp \
+                $(7ZIP_COMMON_DIR)/OutBuffer.cpp \
+                $(7ZIP_COMMON_DIR)/StreamUtils.cpp \
+                $(7ZIP_COMPRESS_DIR)/BitlDecoder.cpp \
+                $(7ZIP_COMPRESS_DIR)/DeflateDecoder.cpp \
+                $(7ZIP_COMPRESS_DIR)/DeflateEncoder.cpp \
+                $(7ZIP_COMPRESS_DIR)/LzOutWindow.cpp \
+                $(7ZIP_COMPRESS_DIR)/ZlibDecoder.cpp \
+                $(7ZIP_COMPRESS_DIR)/ZlibEncoder.cpp \
+                $(SRCDIR)/CPP/Common/MyString.cpp \
+                $(SRCDIR)/CPP/Common/StringConvert.cpp \
+                $(SRCDIR)/CPP/Common/StringToInt.cpp \
+                $(SRCDIR)/CPP/NotWindows/FileIO.cpp \
+                $(SRCDIR)/CPP/NotWindows/MyWindows.cpp \
+                $(SRCDIR)/CPP/Windows/PropVariant.cpp \
+                $(SRCDIR)/deflate7z.cpp
+7ZIP_CXX_TMP := $(7ZIP_CXX_SRC:.cpp=.o)
+7ZIP_CXX_OBJ := $(patsubst $(SRCDIR)/%,$(BLDDIR)/%,$(7ZIP_CXX_TMP))
+
+
+7ZIP_C_SRC :=  $(SRCDIR)/C/Alloc.c \
+               $(SRCDIR)/C/HuffEnc.c \
+               $(SRCDIR)/C/LzFind.c \
+               $(SRCDIR)/C/Sort.c
+7ZIP_C_TMP :=  $(7ZIP_C_SRC:.c=.o)
+7ZIP_C_OBJ := $(patsubst $(SRCDIR)/%,$(BLDDIR)/%,$(7ZIP_C_TMP))
+
+$(BLDDIR)/CPP/7zip/Archive/Common/%.o: $(7ZIP_ARCHIVE_DIR)/Common/%.cpp $(BLDDIR)/.tag
```

Comment:
Hm, personally I would've just used a `.PHONY: builddir` and then `builddir` rule/dependency to always `mkdir -p list of build subdirs`.  If they already exist, the command should be fast anyway.

But I suppose creating a dummy file is fine.  I think it should be ".done" or ".exists" or something instead of ".tag", since that has associated meaning especially with git and might confuse someone.

-[Unknown]

Reply:
The problem is parallel make, its conceivable that it could start compiling before the `mkdir` command is complete causing a failure when the directory does not exist. To my knowledge the best way to deal with this is a sentinel file so that the make rule has an explicit prerequisite on the created file that will only exist after `mkdir` is run.

Reply:
I changed it to `.done`, it doesn't matter to me what the file is called. It could be anything really.

Reply:
Ah, okay, that race condition makes sense.  Thanks.

-[Unknown]